src/url-params.ts

```diff
@@ -49,7 +49,7 @@ export class AnalyticsUrlParams {
 	 * @param queryString URL query string to process
 	 * @return a modified query string that hides processed/consumed parameters if the query string should be modified
 	 */
-	consumeUrlParameters(queryString: string): string | null {
+	consumeUrlParameters(queryString: string): string[] | null {
```

Comment:
You correctly labeled the commit as a major change because it breaks backwards compatibility.
Can we find a way around it?

Reply:
Indeed, I'd prefer not to introduce a breaking change if it's possible.

Reply:
Well the thing is that from what Ive understood of how its built, we could either leave it as it was (not returning the parsed passed device_id, just the query params) but removing the part that overrides the device_id on the client, which we now know its not the intended behaviour. But then for properly initializing the client we should add to it the logic of getting the passed device_id from the cookie that consumeUrlParameters wrote. At least thats another possible way than the one on this commit.

Reply:
@laziob I feel like I'm still missing the motivation behind updating `url-params` file. Let me paraphrase to see if I understand this. 

So we'd like to call `consumeUrlParameters` to parse the passed device identifier and use the first passed identifier as we initialize the client. Then we'd like to link all the device identifiers (passed and/or newly generated) -- which we retrieve from `allDeviceIds`.  

Do I get this right? What's the next problem that we need to solve here?

Reply:
Is the issue on the `setDeviceIds` call in the constructor of `AnalyticsUrlParams`?

Reply:
@gelbal 
> So we'd like to call consumeUrlParameters to parse the passed device identifier and use the first passed identifier as we initialize the client.

Right, so far that is what I think should happen. Which in the end will result in less deviceIds per User since we wont be creating new ones on each init that later on get overriden by the passed one. 

> Then we'd like to link all the device identifiers (passed and/or newly generated) -- which we retrieve from allDeviceIds. 

I dont know exactly when this happens, but by looking at it, I dont think the behaviour of allDeviceIds and linkDevices is affected by this changes.

>Is the issue on the setDeviceIds call in the constructor of AnalyticsUrlParams?

Right, the issue per se in on [this lines](https://github.com/balena-io-modules/analytics-client/blob/ecc900e6699a854dead9a5483e87d00c03d11bfb/src/url-params.ts#L67) , which I commented. 

The other changes are just for returning the parsed passed device_id for using it on init, but that can be changed to any other method we find better or more convenient.

Reply:
```js
if (this.client != null && newCurrentDeviceId != null) {
	this.client.setDeviceId(newCurrentDeviceId);
}
```

Looking at the commented out code, I'd expect this to work as you explain Ezequiel: When we create `AnalyticsUrlParams` without a client set, this line shouldn't run. 

Shall we put aside the changes in `url-params` file and test your other changes? 
I feel like these are sufficient to solve the original issue here.

Reply:
@gelbal if we put aside the changes to `url-params` it wont work, unless we do add on each component away to pass the passed `device_id`.

Right now the order in which the components are using the `analytics-client` is that they first create a client with `createClient()` which by defualt create a `device_id` and an `$identify` event with all the `init_x` properties, and then they run `AnlyticsUrlParams` functions such as `consumeUrlParams`. 

> When we create AnalyticsUrlParams without a client set, this line shouldn't run.

You are right that it wont run, but on the usual implementations like the one I described, theres always a client created. The first issue with the UI was that indeed we were not sending a `client` to the `AnalyticsUrlParams` call and thus we were not setting the proper `device_id` on the client. So following that flow, that line should not run only when no `newCurrentDeviceId` exists, but it should always expect a `client` because otherwise it breaks when there is in fact a `newCurrentDeviceId`.

So the order in which we avoid this kind of issue, is if we first check for a passed `device_id` with `consumeUrlParams`, we save that on cookies or anywhere that works, and only then, we go and create a client with either a passed `newCurrentDeviceId` or with none and we let the `createClient()` create a new `device_id`. 

Thats the reason I made the changes to the `url-params` file. To return both the `newCurrentDeviceId` needed for `createClient()` and the `params.toQueryString()` needed for the URL cleaning.

If I uncomment the commented code, we would still keep this logic of overriding the `device_id` of the client but it would actually never run because on the new needed order of things, `consumeUrlParams` should run before `createClient()` , at least for it to work as I think we intend to.

Reply:
@laziob, thanks for the clarifications. Indeed we need to make changes on the UI repo so the order of initialization is as you explain. Then I suggest doing such modifications on all clients before they start using the new `analytics-client` version.

My general point is: Let's build the ideal solution on `analytics-client` side and enforce the components to follow the expected initialization order. Editing the integration code in the components' repos is in our control. So we could modify the behaviour to follow our best practices. 

Let's assume that we are just rolling out the first version of `analytics-client` and there is no component using it yet, how would you implement this PR? How would you explain to integrate `analytics-client` to gracefully handle the cross domain navigation scenario?

To me it makes sense for a component to utilize `AnalyticsUrlParams` as first thing and parse any incoming parameter that might be required for the `analytics-client` initialization. 

If you agree that this solution would work, I suggest testing it together with the required changes on the UI side. Once we confirm that we solve the issues with the new solution, we would edit the component code to follow our new logic and roll out the new client together.

Reply:
@gelbal Got it! 

Im positive that there might be a more elegant and optimal way to solve for the needed changes than the ones that I suggest here lol, but these changes follow that order of behaviour that I tried to illustrate on my previous comment. 

To go on par with this, I created [this other branch](https://github.com/balena-io/balena-ui/compare/eventLog_changes) on UI to see what it would look like to use this version of `analytics-client` . I think it would help if you could give it a look to see if there is something I missed that could break the implementation. But basically I did exactly that: Change the order in which we run the different functions to comply with these changes.

But either way I will keep on working on this branch to see if  I can come up with a more readable, user friendly way to implement this new logic

Reply:
Cool. Please create a UI PR with that branch Ezequiel. So we could all start commenting on the code and work together towards a solution. I could also start testing the proposed solution in other components like Hub and the marketing website.

I find the current solution pretty nice actually. I will think further if we could come up with a more elegant solution.

Reply:
I wrote this while reviewing another PR but it's just as relevant here:
Maybe deconstructing tuples can help, like
```typescript
const [deviceId, newQuery] = consumeUrlParameters(window.location.search);
```
as described in https://www.carlrippon.com/strongly-typed-destructuring-and-rest-parameters/